Completely replace OpenRA.Test with OpenRA.Utility.exe or Support It

It seems that `OpenRA.Test` does not provide much functionality and I have issues with running it on my system:
`dmcs -t:library -out:OpenRA.Test.dll OpenRA.Test/OpenRA.Game/*.cs -r:mods/common/OpenRA.Mods.Common.dll -r:OpenRA.Game.exe -r:nunit.framework.dll -r:System.dll -r:System.Core.dll -r:System.Data.dll -r:System.Data.DataSetExtensions.dll -r:System.Drawing.dll -r:System.Xml.dll -r:thirdparty/download/ICSharpCode.SharpZipLib.dll -r:thirdparty/download/FuzzyLogicLibrary.dll -r:thirdparty/download/Mono.Nat.dll -r:thirdparty/download/MaxMind.Db.dll -r:thirdparty/download/MaxMind.GeoIP2.dll -r:thirdparty/download/Eluant.dll && nunit-console OpenRA.Test.dll`
results in:
```
OpenRA.Test/OpenRA.Game/ActorInfoTest.cs(67,9): warning CS0219: The variable `i' is assigned but its value is never used
OpenRA.Test/OpenRA.Game/ActorInfoTest.cs(88,9): warning CS0219: The variable `i' is assigned but its value is never used
Compilation succeeded - 2 warning(s)
NUnit version 2.4.8
Copyright (C) 2002-2007 Charlie Poole.
Copyright (C) 2002-2004 James W. Newkirk, Michael C. Two, Alexei A. Vorontsov.
Copyright (C) 2000-2002 Philip Craig.
All Rights Reserved.

Runtime Environment - 
   OS Version: Unix 4.1.3.2
  CLR Version: 4.0.30319.17020 ( 4.0.2 (Stable 4.0.2.5/c99aa0c Mon Jul 13 18:57:35 UTC 2015) )

Symbol file /tmp/nunit20/ShadowCopyCache/8873_635738788612624820/Tests/assembly/shadow/0f6609dd/157c8393_1a1a8a4e_00000001/OpenRA.Test.dll.mdb doesn't match image (null)

Tests run: 0, Failures: 0, Not run: 0, Time: 0.014 seconds


```
That probably is something with my system, but it is another possible dependency / point of failure.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25195330-completely-replace-openra-test-with-openra-utility-exe-or-support-it?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/.gitignore
+++ b/.gitignore
@@ -71,3 +71,7 @@ StyleCopViolations.xml
 # SublimeText
 *.sublime-project
 *.sublime-workspace
+
+# NUnit
+/TestResult.xml
+/lib/

--- a/Makefile
+++ b/Makefile
@@ -6,6 +6,11 @@
 # to compile with development tools, run:
 #   make all [DEBUG=false]
 #
+# to check unit tests (requires NUnit version >= 2.6), run:
+#  make nunit [NUNIT_CONSOLE=<path-to/nunit[2]-console>] [NUNIT_LIBS_PATH=<path-to-libs-dir>] [NUNIT_LIBS=<nunit-libs>]
+#      Use NUNIT_CONSOLE if nunit[2|3]-console was not downloaded by `make dependencies` nor is it not in bin search paths
+#      Use NUNIT_LIBS_PATH if NUnit libs are not in search paths. Include trailing /
+#      Use NUNIT_LIBS if NUnit libs have different names (such as including a prefix or suffix)
 # to check the official mods for erroneous yaml files, run:
 #   make test
 #
@@ -223,12 +228,21 @@ check: utility mods
 	@echo "Checking for code style violations in OpenRA.Test..."
 	@mono --debug OpenRA.Utility.exe ra --check-code-style OpenRA.Test
 
-NUNIT_CONSOLE := $(shell which nunit2-console 2>/dev/null || which nunit3-console 2>/dev/null || which nunit-console 2>/dev/null)
+NUNIT_CONSOLE := $(shell test -f thirdparty/download/nunit-console.exe && echo mono thirdparty/download/nunit-console.exe || \
+	which nunit2-console 2>/dev/null || which nunit3-console 2>/dev/null || which nunit-console 2>/dev/null)
 nunit: test_dll
 	@echo
 	@echo "Checking unit tests..."
-	@if [ "$(NUNIT_CONSOLE)" = "" ] ;then echo 'nunit[2|3]-console not found! Is NUnit installed?'>&2; exit 1; fi
-	@if $(NUNIT_CONSOLE) --help | head -n 1 | grep -E "NUnit version (1|2\.[0-5])";then echo 'NUnit version >= 2.6 required'>&2; exit 1; fi
+	@if [ "$(NUNIT_CONSOLE)" = "" ] ; then \
+		echo 'nunit[2|3]-console not found!'; \
+		echo 'Was "make dependencies" called or is NUnit installed?'>&2; \
+		echo 'see "make help"'; \
+		exit 1; \
+	fi
+	@if $(NUNIT_CONSOLE) --help | head -n 1 | grep -E "NUnit version (1|2\.[0-5])";then \
+		echo 'NUnit version >= 2.6 required'>&2; \
+		exit 1; \
+	fi
 	@$(NUNIT_CONSOLE) $(test_dll_TARGET)
 
 test: utility mods
@@ -466,7 +480,7 @@ help:
 	@echo
 	@echo 'to check unit tests (requires NUnit version >= 2.6), run:'
 	@echo '  make nunit [NUNIT_CONSOLE=<path-to/nunit[2]-console>] [NUNIT_LIBS_PATH=<path-to-libs-dir>] [NUNIT_LIBS=<nunit-libs>]'
-	@echo '     Use NUNIT_CONSOLE if nunit[2|3]-console is not in bin search paths'
+	@echo '     Use NUNIT_CONSOLE if nunit[2|3]-console was not downloaded by `make dependencies` nor is it not in bin search paths'
 	@echo '     Use NUNIT_LIBS_PATH if NUnit libs are not in search paths. Include trailing /'
 	@echo '     Use NUNIT_LIBS if NUnit libs have different names (such as including a prefix or suffix)'
 	@echo

--- a/Makefile
+++ b/Makefile
@@ -41,6 +41,8 @@ CSC         = dmcs
 CSFLAGS     = -nologo -warn:4 -codepage:utf8 -unsafe -warnaserror
 DEFINE      = TRACE
 COMMON_LIBS = System.dll System.Core.dll System.Data.dll System.Data.DataSetExtensions.dll System.Drawing.dll System.Xml.dll thirdparty/download/ICSharpCode.SharpZipLib.dll thirdparty/download/FuzzyLogicLibrary.dll thirdparty/download/Mono.Nat.dll thirdparty/download/MaxMind.Db.dll thirdparty/download/MaxMind.GeoIP2.dll thirdparty/download/Eluant.dll thirdparty/download/SmarIrc4net.dll
+NUNIT_LIBS_PATH :=
+NUNIT_LIBS  := $(NUNIT_LIBS_PATH)nunit.framework.dll
 
 DEBUG = true
 ifeq ($(DEBUG), $(filter $(DEBUG),false no n off 0))
@@ -129,6 +131,16 @@ mod_common_LIBS = $(COMMON_LIBS) $(STD_MOD_LIBS) thirdparty/download/StyleCop.dl
 PROGRAMS += mod_common
 mod_common: $(mod_common_TARGET)
 
+# NUnit testing
+test_dll_SRCS := $(shell find OpenRA.Test/ -iname '*.cs')
+test_dll_TARGET = OpenRA.Test.dll
+test_dll_KIND = library
+test_dll_DEPS = $(game_TARGET)
+test_dll_FLAGS = -warn:1
+test_dll_LIBS = $(COMMON_LIBS) $(game_TARGET) $(NUNIT_LIBS) thirdparty/download/StyleCop.dll thirdparty/download/StyleCop.CSharp.dll thirdparty/download/StyleCop.CSharp.Rules.dll
+PROGRAMS += test_dll
+test_dll: $(test_dll_TARGET)
+
 ##### Official Mods #####
 
 STD_MOD_LIBS	= $(game_TARGET)
@@ -210,6 +222,14 @@ check: utility mods
 	@echo
 	@echo "Checking for code style violations in OpenRA.Test..."
 	@mono --debug OpenRA.Utility.exe ra --check-code-style OpenRA.Test
+
+NUNIT_CONSOLE := $(shell which nunit2-console 2>/dev/null || which nunit3-console 2>/dev/null || which nunit-console 2>/dev/null)
+nunit: test_dll
+	@echo
+	@echo "Checking unit tests..."
+	@if [ "$(NUNIT_CONSOLE)" = "" ] ;then echo 'nunit[2|3]-console not found! Is NUnit installed?'>&2; exit 1; fi
+	@if $(NUNIT_CONSOLE) --help | head -n 1 | grep -E "NUnit version (1|2\.[0-5])";then echo 'NUnit version >= 2.6 required'>&2; exit 1; fi
+	@$(NUNIT_CONSOLE) $(test_dll_TARGET)
 
 test: utility mods
 	@echo
@@ -438,32 +458,38 @@ uninstall:
 	@-$(RM_F) "$(DESTDIR)$(mandir)/man6/openra.6"
 
 help:
-	@echo to compile, run:
-	@echo \ \ make [DEBUG=false]
+	@echo 'to compile, run:'
+	@echo '  make [DEBUG=false]'
 	@echo
-	@echo to compile with development tools, run:
-	@echo \ \ make all [DEBUG=false]
+	@echo 'to compile with development tools, run:'
+	@echo '  make all [DEBUG=false]'
 	@echo
-	@echo to check the official mods for erroneous yaml files, run:
-	@echo \ \ make test
+	@echo 'to check unit tests (requires NUnit version >= 2.6), run:'
+	@echo '  make nunit [NUNIT_CONSOLE=<path-to/nunit[2]-console>] [NUNIT_LIBS_PATH=<path-to-libs-dir>] [NUNIT_LIBS=<nunit-libs>]'
+	@echo '     Use NUNIT_CONSOLE if nunit[2|3]-console is not in bin search paths'
+	@echo '     Use NUNIT_LIBS_PATH if NUnit libs are not in search paths. Include trailing /'
+	@echo '     Use NUNIT_LIBS if NUnit libs have different names (such as including a prefix or suffix)'
 	@echo
-	@echo to generate documentation aimed at modders, run:
-	@echo \ \ make docs
+	@echo 'to check the official mods for erroneous yaml files, run:'
+	@echo '  make test'
 	@echo
-	@echo to install, run:
-	@echo \ \ make \[prefix=/foo\] \[bindir=/bar/bin\] install
+	@echo 'to generate documentation aimed at modders, run:'
+	@echo '  make docs'
 	@echo
-	@echo to install with development tools, run:
-	@echo \ \ make \[prefix=/foo\] \[bindir=/bar/bin\] install-all
+	@echo 'to install, run:'
+	@echo '  make [prefix=/foo] [bindir=/bar/bin] install'
 	@echo
-	@echo to install Linux startup scripts, desktop files and icons
-	@echo \ \ make install-linux-shortcuts [DEBUG=false]
+	@echo 'to install with development tools, run:'
+	@echo '  make [prefix=/foo] [bindir=/bar/bin] install-all'
 	@echo
-	@echo to uninstall, run:
-	@echo \ \ make uninstall
+	@echo 'to install Linux startup scripts, desktop files and icons'
+	@echo '  make install-linux-shortcuts [DEBUG=false]'
 	@echo
-	@echo to start the game, run:
-	@echo \ \ openra
+	@echo 'to uninstall, run:'
+	@echo '  make uninstall'
+	@echo
+	@echo 'to start the game, run:'
+	@echo '  openra'
 
 
 
@@ -474,4 +500,4 @@ help:
 
 .SUFFIXES:
 
-.PHONY: core tools package all mods clean distclean dependencies version $(PROGRAMS)
+.PHONY: core tools package all mods clean distclean dependencies version $(PROGRAMS) nunit

--- a/thirdparty/fetch-thirdparty-deps.sh
+++ b/thirdparty/fetch-thirdparty-deps.sh
@@ -78,6 +78,16 @@ if [ ! -f nunit.framework.dll ]; then
 	rm -rf NUnit
 fi
 
+if [ ! -f nunit-console.exe ]; then
+	echo "Fetching NUnit.Runners from NuGet"
+	get NUnit.Runners 2.6.4
+	cp ./NUnit.Runners/tools/nunit-console.exe .
+	chmod +x nunit-console.exe
+	cp ./NUnit.Runners/tools/nunit-console.exe.config .
+	cp -R ./NUnit.Runners/tools/lib .
+	rm -rf NUnit.Runners
+fi
+
 if [ ! -f Mono.Nat.dll ]; then
 	echo "Fetching Mono.Nat from NuGet"
 	get Mono.Nat 1.2.21

